guidance/library/_sequences.py

```diff
@@ -0,0 +1,52 @@
+from typing import Union
+
+from .._grammar import Join, select
+from .._guidance import guidance
+from ._optional import optional
+
+
+@guidance(stateless=True)
+def exactly_n_repeats(model, value, n_repeats: int):
+    assert n_repeats >= 0, f"Must have n_repeats>=0 ({n_repeats})"
+    return model + Join([value] * n_repeats)
+
+
+@guidance(stateless=True)
+def at_most_n_repeats(model, value, n_repeats: int):
+    assert n_repeats >= 0
+
+    return model + exactly_n_repeats(optional(value), n_repeats)
+
+
+@guidance(stateless=True)
+def sequence(
+    model, value, min_length: Union[int, None] = None, max_length: Union[int, None] = None
+):
+    assert min_length is None or min_length >= 0
+    assert max_length is None or max_length >= 0
+
+    if min_length is not None and max_length is not None:
+        assert max_length >= min_length
+        model += exactly_n_repeats(value=value, n_repeats=min_length)
+        model += at_most_n_repeats(value=value, n_repeats=(max_length - min_length))
+    elif min_length is not None:
+        model += exactly_n_repeats(value=value, n_repeats=min_length)
+        model += select([optional(value)], recurse=True)
```

Comment:
Can you reuse zero_or_more here to reduce code duplication?

Reply:
I had actually been planning on reimplementing those two functions in terms of `sequence()`. Of course, _something_ isn't quite right, and the errors I'm getting are *very* Pythonic.

Reply:
Oh fair enough. Sorry about your python 😂

Reply:
There is one test involving the mocked model which is interfering with `zero_or_more()` being a synonym for `sequence()`. It's related to the max_tokens argument on gen, so not a core part of this PR, though.

Reply:
Would you open an issue if it feels worthy? :)

Reply:
Done. It could also be something strange in the way the mock model is working. But it would be good to run down.